openedx/core/djangoapps/catalog/utils.py

```diff
@@ -252,36 +252,44 @@ def get_pseudo_session_for_entitlement(entitlement):
     Returns None if there are no sessions for that course.
     """
     sessions_for_course = get_course_runs_for_course(entitlement.course_uuid)
-    available_sessions = get_fulfillable_course_runs_for_entitlement(entitlement, sessions_for_course)
+    available_sessions = get_fulfillable_course_runs_for_entitlement(entitlement, sessions_for_course,
+                                                                     entitlements=entitlements)
     if available_sessions:
         return available_sessions[0]
     if sessions_for_course:
         return sessions_for_course[0]
     return None
 
 
-def get_visible_sessions_for_entitlement(entitlement):
+def get_visible_sessions_for_entitlement(entitlement, entitlements=None):
     """
     Takes an entitlement object and returns the course runs that a user can currently enroll in.
     """
     sessions_for_course = get_course_runs_for_course(entitlement.course_uuid)
-    return get_fulfillable_course_runs_for_entitlement(entitlement, sessions_for_course)
+    return get_fulfillable_course_runs_for_entitlement(entitlement, sessions_for_course, entitlements=entitlements)
 
 
-def get_fulfillable_course_runs_for_entitlement(entitlement, course_runs):
+def get_fulfillable_course_runs_for_entitlement(entitlement, course_runs, entitlements=None):
     """
     Takes a list of course runs and returns only the course runs that:
 
     1) Are currently running or in the future
     2) A user can enroll in
     3) A user can upgrade in
     4) Are published
+    5) Are not enrolled in by a user's other entitlements
 
     These are the only sessions that can be selected for an entitlement.
     """
 
     enrollable_sessions = []
 
+    enrolled_sessions = []
+    if entitlements:
+        for course_entitlement in entitlements:
+            if course_entitlement.enrollment_course_run:
+                enrolled_sessions.append(str(course_entitlement.enrollment_course_run.course_id))
```

Comment:
As I mentioned in person, this feels brittle to me. If what we care about is the question "is the user enrolled in this course run?", going from entitlements->runs seems indirect and encodes assumptions about possible ways the user can be in a run (i.e. assumes only entitlements do that). That's probably true today, but still.

Is there a performant way of asking "what is the set of course run IDs that the user is enrolled in" and just get that info directly rather than through entitlements?

Reply:
Yeah there definitely should be a way to make that query, I'll look into doing a query using the `student_courseenrollment` table, there should be something like `SELECT course_id FROM student_courseenrollment WHERE user_id=? AND mode='verified';`